### DISCUSSION TITLE
ISR: Should client-side navigation with next/link trigger a page to revalidate?

### DISCUSSION THREAD

**Discussion by sdjnes** [vercel/next.js#42685] [Help]
Hello,

In discussions/issues, I've found information that ISR revalidation is only triggered on document load of a Next.js page (i.e. directly visiting the URL, or refreshing the page). For many websites and web apps, users land on a homepage and navigate to other pages with links rather than directly going to subpages via a URL.

If page content is pulled from an external service, and changes over time, subpages will show stale content until a refresh or direct visit is made to the page.

In [this discussion](https://github.com/vercel/next.js/discussions/40521), it seems that client-side navigation does/did trigger revalidation of pages, however, the examples given there don't work for me as of today. 

The [ISR documentation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration) is a little ambiguous with the wording: "The only way to invalidate the cache is from someone visiting that page after the minute has passed.". I incorrectly assumed "visiting" would also include client-side navigation, as it's one of those things that I use in the project without much thought and is the recommended way. I'd guess the majority of Next.js projects also use client-side navigation with next/link.

-- 

# Possible solutions
## On-demand revalidation
This seems to be the goto stratgey and I believe it works well in some use cases. For example, when content changes in the external service, call an API route in the Next.js application `/api/revalidate?path={path}` which revalidates that specific path.

However, if there is content that serves multiple pages (e.g. header, footer), this would require many calls to the `revalidate()` function in the API route (as there is no `revalidateAllPaths()`). Deriving the paths can also be an issue, in my case this would require manual work to keep these paths up to date if new non-dynamic routes are added to the repo.

Using Vercel, it would be easy to hit the timeout limits for the serverless functions when revalidating many pages.

## Use `a` tags as opposed to `next/link`
I think this is likely the best strategy, with some sacrifice to performance. 

## (Future?) `getStaticProps` in [layouts](https://nextjs.org/docs/basic-features/layouts#data-fetching)
If layouts were to support `getStaticProps` in the future, this might solve the above issue of multiple calls to `revalidate()` for paths, as we could revalidate common page components independently.

-- 

I guess my questions are:

* Does anybody have client-side navigation working with next/link client-side navigation?
* Is the definitive answer that ISR doesn't work this way, and what's the reasoning behind it?
* Could the documentation be updated to reflect this?
* Is there any plans for `getStaticProps` to be supports in layouts?

**Comment by icyJoseph** [comment]
I made an app where pages with `getStaticProps` return props and revalidate 10. As part of props I send a timestamp and then I show that on the page. With SPA transitions, using `next/link` I see the revalidated data. 

Tested even with middleware activated.

What version of Next.js are you using? Can you share a minimal repro where it doesn't work as expected?

**Reply by sdjnes** [reply]
I'm on 12.2.5, but I've also seen the issue on 12.3.0.

Here's a minimal reproduction: https://stackblitz.com/edit/nextjs-cvdz67. This is a slightly modified version of the example given [here](https://github.com/vercel/next.js/discussions/40521)

After running `npm run build && npm run start`, I'm able to click back-and-forth between the About Us and Homepage using the next/link and despite the revalidate being set to `10`, I see the following:

![Screenshot 2022-11-10 at 11 31 53](https://user-images.githubusercontent.com/6553059/201067909-fe4f277e-9ea1-400b-afaf-5807384a3370.png)

Refreshing the About Us page triggers the revalidate after 10 seconds as expected. What I have noticed is if I'm on a different tab for a little while and return, it *does* trigger the revalidate.

Seen in both Chrome (Version 107.0.5304.110) and Firefox Dev Edition (v107.0b9)

**Reply by icyJoseph** [reply]
Hi,

I am only able to sort of reproduce this on Stackblitz, and that's because there's a service worker at play there, which caches responses. Before going deep into that, are you seeing this on your dev environment? Do you have a Service Worker (PWA)?

**Reply by sdjnes** [reply]
Good spot on the Stackblitz, I guess that rules out the reproducability there. No service workers either.

It's working locally, but I didn't catch it before... basically I have 2 pages with the next/link in the same place and if I do not re-hover it, the page will not revalidate. That's likely a different issue, so the good news is it's working locally

One thing I didn't mention is that I'm deploying the project to Vercel, where ISR revalidate with next/link simply does not work on Preview or Production deployments under any circumstances. Could this be a Vercel specific issue?

**Reply by icyJoseph** [reply]
Mmm unlikely, but I'd debug this by opening the network tab and making sure `.json` requests are coming through, and if they do validate the data.

The next/link detects when it is on the viewport, and unless told not to prefetch via props, it'll make a call to the server for JSON props. For GSSP pages it wait for the actual click. 

Even if you disable prefetch, the next/link prefetches JSON props on hover, and again, for GSSP pages it waits for the click. There was a bug a couple of versions back where GSSP was prefetched before the click and never again, but IIRC 12.3.x fixed it.

For all cases next/link prefetches the JavaScript that runs the page.

JSON props with revalidate have a short cache lifetime, this is something you might want to debug too.

There's also the thing that happens with `notFound: true` which is cached for a whole year, so you ought to return a revalidate prop too.

**Reply by sdjnes** [reply]
I can see the JSON requests coming through fine when the link appears in the view port and on hover of the link, but the data of the JSON is always stale unless I refresh the page. If I visit the JSON resource URL directly in my browser and refresh a couple of times, it's revalidated. But it seems any client side requests for that JSON resource does no revalidate it.

Here's a snapshot of the headers for one of the requests:

```
age: 239
cache-control: public, max-age=0, must-revalidate
x-vercel-cache: STALE
```

I believe the cache-control headers are correct for my configuration (`revalidate: true`), but it also appears that Vercel knows the cache for the JSON resource is stale, but until I refresh the page, the `age` keeps increasing and the data does not get revalidated.

**Reply by icyJoseph** [reply]
And this also happens in 12.3.0 right? I do recall there was a bug in 12.2.5 and 12.2.6 but I am not sure if it was related at all. 

Locally, under `next build && next start` it works fine right?

**Reply by sdjnes** [reply]
12.2.5 and 12.3.0 locally it works as expected. The project was on 12.2.5 originally, but the upgrade to 12.3.0 required minimal code changes so I upgraded it.

I deployed 12.3.0, and the revalidation worked exactly once. I updated the page in the CMS, and visiting it via client side navigation eventually showed the new content, however, any other page I updated (including the original one) all showed stale content on the frontend.

Interestingly, the network request for the JSON resources do have the new content in the response, but the actualy frontend render is showing stale data.

EDIT: Can't seem to pin it down. Sometimes the JSON resource has the latest data, sometimes not. In some cases, the JSON resource is showing old data, but because I refreshed the page itself, the frontend render shows the latest data and not the data that was returned in the JSON.

**Reply by dubbajubba** [reply]
@sdjnes , I am having very similar issues with my project. Hopefully someone can shed some light and get us a resolution!

**Reply by sdjnes** [reply]
Good to know I'm not alone @austin-triiistudio!

I'm planning to deploy a minimal reproducible example to Vercel but haven't had chance yet. The project I'm seeing this issue in has some additional logic in the `_app` file so I want to make sure that isn't interfering with anything.

**Reply by sdjnes** [reply]
Here's the repo and deployment: 

https://github.com/sdjnes/vercel-next-isr-test
https://vercel-next-isr-test.vercel.app/

Of note, I've set Next.js to `12.3.1` and React to `17` as this is what I'm running the project where I see the issue.

Here's a video of me clicking back-and-forth with `next/link`. Notice the `getStaticProps` timestamp not updating until I do a page refresh to trigger the revalidation, and a subsequent refresh to see the new content.

I have also created a preview deployment for branch `next-13` in that repo, which uses the Next.js and React version defaults from `create-next-app` (deployment here: https://vercel-next-isr-test-nsnaodexd-sjones-inpostcouk.vercel.app/) and the issue is also seen for me.

https://user-images.githubusercontent.com/6553059/202306179-76fd0d07-7664-49c1-af1d-952300e8f4e5.mov

**Comment by IronDdtRIiGh** [comment]
Sould no revalidate should rerssintsssinte., u. ^%<@#_
exut#_%_<

**Comment by krsntn** [comment]
@sdjnes , has the issue been resolved yet? I am facing the same problem with my project which is using next `12.3.4` and react `18.2.0`. The project runs without any issues when I build and start it locally using `next build && next start`, but I am encountering the same issue when deploying it on Vercel.

**Reply by sdjnes** [reply]
It hasn't been resolved yet as far as I'm aware. I opened a GitHub issue after confirming with Vercel that it was indeed an issue (https://github.com/vercel/next.js/issues/43289) and was told they would put any updates there
